### ISSUE TITLE
Sudden database corruption in v1.0.174

### ISSUE THREAD

**Issue by heinrich-ulbricht** [laurent22/joplin#2129]
## Environment

Joplin version: 1.0.174
Platform: Windows 10
OS specifcs: -

## Steps To Reproduce

Don't know.

I left the computer for about an hour, came back, continued editing. I could still edit the note (so the edit area allowed input) but the preview was stuck and didn't update and no buttons could be clicked anymore. Closed, reopened - boom - error message (see below). Now Joplin only shows this error message and immediately closes. Can I do something?

(I inserted an image at some point before this situation happened. Might be related, might be not.)

## Logfile

This was the error dialog shown upon restarting Joplin:

```
[Window Title]
Joplin

[Content]
Fatal error:

Error: SQLITE_CORRUPT: database disk image is malformed: CREATE TABLE folders (id TEXT PRIMARY KEY, title TEXT NOT NULL DEFAULT "", created_time INT NOT NULL, updated_time INT NOT NULL);

Error: Error: SQLITE_CORRUPT: database disk image is malformed: CREATE TABLE folders (id TEXT PRIMARY KEY, title TEXT NOT NULL DEFAULT "", created_time INT NOT NULL, updated_time INT NOT NULL);
    at DatabaseDriverNode.sqliteErrorToJsError (C:\Users\myuser\AppData\Local\Temp\nsa8121.tmp\app\resources\app.asar\lib\database-driver-node.js:21:16)
    at JoplinDatabase.sqliteErrorToJsError (C:\Users\myuser\AppData\Local\Temp\nsa8121.tmp\app\resources\app.asar\lib\database.js:22:24)
    at JoplinDatabase.tryCall (C:\Users\myuser\AppData\Local\Temp\nsa8121.tmp\app\resources\app.asar\lib\database.js:85:17)
    at process._tickCallback (internal/process/next_tick.js:68:7)

[OK]
```

Log entry:
```
2019-11-28 12:39:43: "Scheduling sync operation..."
2019-11-28 12:39:45: "Scheduling sync operation..."
2019-11-28 12:39:48: "Scheduling sync operation..."
2019-11-28 12:39:53: "SearchEngine: Updating FTS table..."
2019-11-28 12:39:54: "SearchEngine: Updated FTS table in 35ms. Inserted: 1. Deleted: 0"
2019-11-28 12:40:03: "Scheduling sync operation..."
2019-11-28 12:40:45: "Preparing scheduled sync"
2019-11-28 12:40:45: "Starting scheduled sync"
2019-11-28 12:40:45: "Could not run background sync:"
2019-11-28 12:40:45: "Error: Error: SQLITE_CORRUPT: database disk image is malformed: SELECT count(*) as total FROM `master_keys`
Code: SQLITE_CORRUPT
Error: Error: SQLITE_CORRUPT: database disk image is malformed: SELECT count(*) as total FROM `master_keys`
    at DatabaseDriverNode.sqliteErrorToJsError (C:\Users\myuser\AppData\Local\Temp\nsb50FA.tmp\app\resources\app.asar\lib\database-driver-node.js:21:16)
    at JoplinDatabase.sqliteErrorToJsError (C:\Users\myuser\AppData\Local\Temp\nsb50FA.tmp\app\resources\app.asar\lib\database.js:22:24)
    at JoplinDatabase.tryCall (C:\Users\myuser\AppData\Local\Temp\nsb50FA.tmp\app\resources\app.asar\lib\database.js:85:17)
    at process._tickCallback (internal/process/next_tick.js:68:7)"
2019-11-28 12:45:45: "Running background sync on timer..."
2019-11-28 12:45:45: "Scheduling sync operation..."
2019-11-28 12:45:45: "Preparing scheduled sync"
2019-11-28 12:45:45: "Starting scheduled sync"
2019-11-28 12:45:45: "Synchronisation is already in progress. State: in_progress"
2019-11-28 12:47:43: "RevisionService::maintenance: Starting..."
2019-11-28 12:47:43: "RevisionService::maintenance: Service is enabled"
2019-11-28 12:47:43: "RevisionService::collectRevisions:", "Error: Error: SQLITE_CORRUPT: database disk image is malformed: 
					SELECT id, item_id, type, before_change_item
					FROM item_changes
					WHERE item_type = ?
					AND source != ?
					AND source != ?
					AND id > ?
					ORDER BY id ASC
					LIMIT 10
				: 1,2,2,2125
Code: SQLITE_CORRUPT
Error: Error: SQLITE_CORRUPT: database disk image is malformed: 
					SELECT id, item_id, type, before_change_item
					FROM item_changes
					WHERE item_type = ?
					AND source != ?
					AND source != ?
					AND id > ?
					ORDER BY id ASC
					LIMIT 10
				: 1,2,2,2125
    at DatabaseDriverNode.sqliteErrorToJsError (C:\Users\myuser\AppData\Local\Temp\nsb50FA.tmp\app\resources\app.asar\lib\database-driver-node.js:21:16)
    at JoplinDatabase.sqliteErrorToJsError (C:\Users\myuser\AppData\Local\Temp\nsb50FA.tmp\app\resources\app.asar\lib\database.js:22:24)
    at JoplinDatabase.tryCall (C:\Users\myuser\AppData\Local\Temp\nsb50FA.tmp\app\resources\app.asar\lib\database.js:85:17)
    at process._tickCallback (internal/process/next_tick.js:68:7)"
2019-11-28 12:50:45: "Running background sync on timer..."
```

Note: I cannot follow the steps outlined here: _https://joplinapp.org/debugging/_ - there is no _joplin-desktop_ folder on my system. I use the portable version - should _flags.txt_ go into the root of the portable folder?

**Comment by laurent22** [comment]
Database corruption can happen for a number of reasons unrelated to Joplin, and unfortunately when it happens there's nothing to do. All you can do is clear your profile and restore from your latest backup.

**Comment by heinrich-ulbricht** [comment]
Hm. After restoring from a backup it happened again almost immediately :( Might be something on my system, might be not. If this is a common scenario I'd like to see a restore or proactive monitoring built-in.

**Comment by laurent22** [comment]
This is extremely uncommon, and the app cannot directly corrupt the database (even if we wanted to) without an external factor, such as failing hard drive, driver problems, etc.

**Comment by laurent22** [comment]
If it happens immediately after you restore your backup, it's possible that your backup file is also corrupted. Can you open it with something like SQlite Browser?

**Comment by heinrich-ulbricht** [comment]
I restored to a previous version of _database.sqlite_ which I could open with SQLiteBrowser. I replaced the corrupted one (that SQLiteBrowser couldn't open) with the working one and Joplin opened fine. I started putting some text in and after a very short time (a minute?) it again went into "can type, but actually nothing works" mode. After restarting Joplin I got the same error as before. Yes, something could be failing on my machine or some other software might interfer. (I hope not.) Who knows :-/
